src/format/Token.cc

```diff
@@ -480,9 +487,14 @@ Format::Token::parse(const char *def, Quoting *quoting)
 
     case LFT_NOTE:
 
+    case LFT_PROXY_PROTOCOL_RECEIVED_HEADER:
+
         if (data.string) {
             char *header = data.string;
-            char *cp = strchr(header, ':');
+            // HTTP header field names cannot have ':' while
+            // PROXY protocol pseudo headers start with it.
```

Comment:
> Implementing this change request would be wrong because proxyPseudoHeader is not set to true for all pseudo headers. It is only set to true for PROXY protocol pseudo headers. The code using this variable depends on this.

Ah the lack of brackets fooled my eyes there.

> 
> It is true that the PROXY protocol specs do not use the "pseudo header" terminology (yet?), but that does not stop us from using that (general) terminology. The term matches the use case perfectly, there is no good replacement within the PROXY protocol specs, and the future HTTP/2 support would only benefit from this naming.
> 

I don't see how HTTP/2 benefits from the HTTP pseudo-headers being called PROXY-protocol pseudo-header.


> N.B. The term "pseudo header" [predates](https://en.wikipedia.org/wiki/User_Datagram_Protocol#IPv6_Pseudo_Header) HTTP/2 by many years. It is (correctly) used by HTTP/2, but it is certainly not HTTP/2-specific.

So where did Eduard get this definition of pseudo-header as a lower case string starting with a ':' if not HTTP/2 documents?

Reply:
So essentially you took inspiration and based a pseudo-header definition on the HTTP/2 design. As I said.

The problem I have with this code is that a) the "pseudo headers" is your own invention, but the documentation makes it sound like they are a thing defined by PROXY protocol. And b) that the variable name reinforces that false implication.

We do not actually gain anything by mentioning PROXY protocol here. We *loose* the ability to log pseudo-header fields by preventing them generically. While also forcing extra work on dev trying to add other protocols which share this ":name" pseudo-header design (aka, HTTP/2 and HTTP/3). What one would call a net-negative.

Reply:
@eduard-bagdasaryan, I suggest applying the changes @yadij insists on in this change request, simply to avoid wasting more time on this minor issue:

* The source comment content is just not that important. Amos sees a minor implication that you think is not there. That argument is nearly impossible to resolve using logic, even if all sides are perfectly reasonable (perfectly reasonable folks would not have such arguments to start with, but nobody is perfect).
* As for the "restriction" that the proxyPseudoHeader variable is only true for PROXY protocol, I suggest removing the `LFT_PROXY_PROTOCOL_RECEIVED_HEADER` condition from the variable initialization code and renaming the variable to pseudoHeader. To avoid creating a bug where something like `:Accept-Encoding` is misinterpreted as an HTTP `Accept-Encoding` header, add another `if` statement to throw an error if a pseudo header is used for the `type` value other than `LFT_PROXY_PROTOCOL_RECEIVED_HEADER`. These modifications would make the new pseudo header configuration support a bit safer AFAICT.

Hopefully, the above adjustments will address this change request.

Reply:
This change broke parsing of integer-based TLV value types. I have a fix at https://github.com/measurement-factory/squid/commit/39b68165ad34dd95647e0caaa990aa23254cd43b

While fixing that bug, I also notice that our detection and diagnostics for certain likely configuration mistakes is sloppy and added https://github.com/measurement-factory/squid/commit/98adf59e597f95969fbac1c36c9a68fd46cf653f

Please review/retest/adjust and integrate into this PR.

Reply:
For the record: Eduard has added both fixes as of branch commit 878714dce37300225b833b719e5f072143183e15. Determining whether this particular change request has been addressed remains on @yadij side.